src/dials/command_line/combine_experiments.py

```diff
@@ -745,6 +745,8 @@ def save_in_batches(
 
         # cluster the resulting experiments if requested
         if params.clustering.use and len(experiments) > 1:
+            if params.clustering.max_clusters == 0:
+                sys.exit("Error: max_clusters must be None or >0")
```

Comment:
This seems reasonable. Is there any argument for using `sys.exit` instead of `ValueError`, which would give a clearer traceback? For example is there a chance the latter would be caught somewhere by mistake?

Reply:
Yes! A stack trace here would be seen by the user. IMO a user should never be expected to have to figure out a stack trace (I say User STacktrace Always Bad), so we should aspire to avoid them in cases where it is specific feedback to the user about something that might be in their control. From the POV of the code it's an unrecoverable error (unless we want to guess what the user meant) and specifically _here_ we are in `dials.command_line.<something>.<function handling user invocation>` and not somewhere in e.g. the algorithms library (where a `ValueError` might be appropriate) so terminating is appropriate. This does admittedly get a little fuzzy from time-to-time as we don't have great separation between validating input and things just going plain wrong (and not a lot of Exception-level categorisation) but still I think something to aim towards.

In days of yore we used `Sorry` for this purpose of "We have an unrecoverable error and cannot fix and want to terminate with a specific message for the user" - however, we came to the conclusion that functionally this was identical to the `SystemExit` exception (including when and how we would want to catch) so try to use `sys.exit` directly nowadays.

I also noticed the other day that `dials.integrate` specifically does https://github.com/dials/dials/blob/995d6c2e2ffd4e43162961d2e309a9b0522e2bcc/src/dials/command_line/integrate.py#L728-L729

which should be applauded as the best kind of correct for being _technically_ correct in never showing the user a stack trace, but isn't really following the spirit of the aspiration (and is not very useful when a user does send us an actual problem report and we have to guess where it came from)

Reply:
Thank you, that was insightful. I learned that exceptions are generally preferable over system exit, but I can see why it might have its use in this case.

Reply:
In principle I think the two choices `sys.exit()` and `raise Sorry` are supposed to be different when using mpi multiprocessing. This script `good.py`:

```
from mpi4py import MPI
from dials.util import Sorry

comm = MPI.COMM_WORLD
if comm.rank == 4: raise Sorry('rank 4 is unhappy')
MPI.COMM_WORLD.barrier()
print('ok')
```
should crash (as desired) when called with `mpirun -n 5 python good.py` whereas this `bad.py`:

```
from mpi4py import MPI
import sys

comm = MPI.COMM_WORLD
if comm.rank == 4: sys.exit('rank 4 is unhappy')
MPI.COMM_WORLD.barrier()
print('ok')
```
should confusingly hang forever.

In practice they currently both appear to hang forever due to faulty OpenMP signal handling as discussed here: https://github.com/chainer/chainermn/issues/236. A segfault on one rank is still handled correctly as in `actuallygood.py`:

```
from mpi4py import MPI

comm = MPI.COMM_WORLD
if comm.rank==4: import ctypes;ctypes.string_at(0)
MPI.COMM_WORLD.barrier()
print('ok')
```

...not that I'm advocating for showing segfaults to users. However maybe this is something to think about, I've definitely been bitten by this before when an MPI rank hits a `sys.exit`.

Reply:
@dwpaley This makes sense. Now that you mention that, I think that explains one of my bigger problemsI've encounter when processing current data. When calling `dials.refine` from cctbx's `ensemble_refinement` worker, if the number if ranks per node is too high, a `DivisionByZero` error is called due to an apparent lack of memory. If my memory does not fail me, this error is raised on multiple nodes and causes the whole process to hang instead of failing. This was especially painful to debug, as lack of memory incentified me to use more nodes, which just ate more NERSC resources without solving the issue.

Reply:
This is because dials.refine and dials.combine_experiments aren't "mpi safe".  Meaning since we invoke them using srun they are called by each mpi rank instead of by just 1 mpi rank.  That can be easily fixed by adding an mpi safe python script for ensemble refinement that would look something like this:

```
if rank == 0:
  call combine and refine
call mpi integrate
```

Something we should do before our next round of experiments.

Reply:
Here's a little experiment to fix the faulty behavior I mentioned above: https://github.com/cctbx/cctbx_project/compare/master...mpi_excepthook If this is adopted then when running under MPI control it really will be better to use `Sorry` because it hits `sys.excepthook`.

Reply:
MPI can't handle the underlying process just outright crashing?

Reply:
The behavior I've seen is that if a rank raises a traceback (or Sorry), then MPI shuts down all other ranks and the program exit.  If a rank calls sys.exit instead, that rank shuts down and the program indefinitely hangs while all the other ranks are still active.  Am I close @dwpaley?

Reply:
This may be MPI implementation dependent?  Like mpich vs. openmpi.

Reply:
Tried that, but I don't think there's a mechanism for the registered function to know if we're exiting successfully or unsuccessfully. This seems better:

```
from mpi4py import MPI
import sys

def mpisafe(func):
  def wrapped_func(*args, **kwargs):
    try:
      return func(*args, **kwargs)
    except Exception:
      sys.excepthook(*sys.exc_info())
      MPI.COMM_WORLD.Abort(1)
    except SystemExit as e:
      if e.code:
        sys.stderr.write(e.code + '\n')
      MPI.COMM_WORLD.Abort(1)
  return wrapped_func

def do_work():
  comm = MPI.COMM_WORLD
  if comm.rank==4: sys.exit('something is wrong')
  return('ok')

@mpisafe
def run():
  print(do_work())
  MPI.COMM_WORLD.barrier()
  print('done')

if __name__ == '__main__':
  run()
```

and the decorator could be defined in `libtbx.mpi4py`.

Reply:
Yep, catching to do cleanup generally seems the safest. Although if it's just cleanup (this presumably being cases where you actually do want a full abort) then
```python
try:
    call_function()
except BaseException:
    MPI.COMM_WORLD.Abort(1)
    raise
```
does the abort, but also propagates the unhandled exception, and doesn't swallow the SystemExit message and return code. I guess there also be a way to query if you closed the MPI session cleanly and to abort in an `atexit`. The documentation also does seems to distinguish between main and non-main threads, but that might only be if explicitly using thread support.

Reply:
Unfortunately if you call Abort then you never hit raise. I don't immediately see a way to do both. Hence manually calling excepthook or printing the exit message.

That means this approach isn't compatible with propagating errors up to the caller (since we have to choose either raise or Abort). So the name `@mpisafe` is not really appropriate, it's more like `@mpi_abort_on_exception` to clarify this can only be applied in a place where exception means failure.